src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -215,6 +213,67 @@ private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, Co
             return false;
         }
 
+        // We should remove Razor design time helpers from C#'s completion list. If the current identifier being targeted does not start with a double
+        // underscore, we trim out all items starting with "__" from the completion list. If the current identifier does start with a double underscore
+        // (e.g. "__ab[||]"), we only trim out common design time helpers from the completion list.
+        private SumType<CompletionItem[], CompletionList>? RemoveDesignTimeItems(
+            SumType<CompletionItem[], CompletionList> completionResult,
+            LSPDocumentSnapshot documentSnapshot,
+            TextExtent? wordExtent)
+        {
+            var result = completionResult.Match<SumType<CompletionItem[], CompletionList>?>(
+                items =>
+                {
+                    items = RemoveDesignTimeItems(items, documentSnapshot, wordExtent);
+                    return items;
+                },
+                list =>
+                {
+                    list.Items = RemoveDesignTimeItems(list.Items, documentSnapshot, wordExtent);
+                    return list;
+                });
+
+            return result;
+
+            static CompletionItem[] RemoveDesignTimeItems(CompletionItem[] items, LSPDocumentSnapshot documentSnapshot, TextExtent? wordExtent)
```

Comment:
lol it's interesting having the nested nested local functions but I don't have a better recommendation. Overall love how much clearer the flow is.